### PR TITLE
kubectl wait: wire generic context

### LINKED ISSUE #1219: kubectl wait timeout argument is poorly documented and ill-suited to waiting on multiple resources
This is just a re-submit of https://github.com/kubernetes/kubectl/issues/754 which, despite being confirmed & assigned, was closed as stale without any fix.

**What happened**:

Run `kubectl wait` with a selector matching more than one resource and a timeout

**What you expected to happen**:

The timeout should apply to the wait command, not to the individual resources.

With the timeout applying to resources sequentially, it makes waiting on more than one resource with any kind of timeout basically unusable.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a deployment scaled to 2 or more replicas, and a label that can be used to match it
2. Run: `kubectl wait pod --selector=... --timeout=30s`
3. Observe that this runs for N*30s, where N is the number of pods

**Anything else we need to know?**:

cc @eranreshef the original reporter and @JabusKotze who assigned the prior issue to themselves

### PR DESCRIPTION
**Pull request by ardaguclu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Currently `kubectl wait` does not rely on the value in `--timeout` flag when there are multiple resources. This PR wires generic context to all resources and when the given timeout occurs, all processes will be cancelled.

In addition to that, while working on this issue, I noticed that `kubectl wait` integration test always passing, this PR also fixes that.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubectl/issues/1219

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by ardaguclu** [comment]
/triage accepted
/priority backlog
/hold 
until review

/cc @mpuckett159

File: test/cmd/wait.sh [hunk 1]
```diff
@@ -56,6 +54,43 @@ run_wait_tests() {
     kube::test::if_has_string "${output_message}" 'test-1 condition met'
     kube::test::if_has_string "${output_message}" 'test-2 condition met'
 
+    # create test data to test timeout error is occurred in correct time
+    kubectl apply -f - <<EOF
```
Comment (ardaguclu) [inline test/cmd/wait.sh:58]: This test is basically copied from this https://github.com/kubernetes/kubectl/issues/1219#issuecomment-1155501551

File: staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go [hunk 2]
```diff
@@ -386,9 +386,14 @@ func IsDeleted(info *resource.Info, o *WaitOptions) (runtime.Object, bool, error
 		return false, nil
 	}
 
+	intrCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 	intr := interrupt.New(nil, cancel)
 	err := intr.Run(func() error {
-		_, err := watchtools.UntilWithSync(ctx, lw, &unstructured.Unstructured{}, preconditionFunc, Wait{errOut: o.ErrOut}.IsDeleted)
+		_, err := watchtools.UntilWithSync(intrCtx, lw, &unstructured.Unstructured{}, preconditionFunc, Wait{errOut: o.ErrOut}.IsDeleted)
+		if errors.Is(err, context.DeadlineExceeded) {
```
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:394]: I assume this will only work when https://github.com/kubernetes/kubernetes/pull/114578 lands?

**Review by soltysh — APPROVED** [review]
/lgtm
/hold cancel

File: staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go (hunk 2 above)
Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:394]: unfortunately, yes, it is correct

**Pull request merged**